elastica/boundary_conditions.py

```diff
@@ -1,70 +1,135 @@
 
 import numpy as np
-from elastica._rotations import _get_rotation_matrix
+
+from abc import ABC, abstractmethod
 
 import numba
 from numba import njit
 
+from elastica._rotations import _get_rotation_matrix
+from elastica.rod import RodBase
+
 
-class FreeRod:
-    """
-    This is the base class for displacement boundary conditions. It applies no constraints or displacements to the rod.
+class ConstraintBase(ABC):
+    """Base class for constraint and boundary condition implementation.
 
     Note
     ----
-    Every new displacement boundary condition class must be
-    derived from FreeRod class.
-    """
+    Constraint class must inherit BaseConstraint class.
 
-    def __init__(self):
-        """
-        Free rod has no input parameters.
-        """
-        pass
+        Attributes
+        -------
+        system : RodBase
+        node_indices : None or numpy.ndarray
+        element_indices : None or numpy.ndarray
 
-    def constrain_values(self, rod, time):
+    """
+
+    _system: Type[RodBase]
+    _constrained_position_idx: np.ndarray
+    _constrained_director_idx: np.ndarray
+
+    def __init__(self, *args, **kwargs):
+        """Initialize boundary condition"""
+        try:
+            self._system = kwargs["_system"]
+            self._constrained_position_idx = np.array(
+                kwargs.get("constrained_position_idx", []), dtype=int
+            )
+            self._constrained_director_idx = np.array(
+                kwargs.get("constrained_director_idx", []), dtype=int
+            )
+        except KeyError:
+            raise KeyError(
+                "Please use simulator.constrain(...).using(...) syntax to establish constraint."
+            )
+
+    @property
+    def system(self) -> Type[RodBase]:
+        """get system (rod or rigid body) reference"""
+        return self._system
+
+    @property
+    def constrained_position_idx(self) -> Optional[np.ndarray]:
+        """get position-indices passed to "using" """
+        # TODO: This should be immutable somehow
+        return self._constrained_position_idx
+
+    @property
+    def constrained_director_idx(self) -> Optional[np.ndarray]:
+        """get director-indices passed to "using" """
+        # TODO: This should be immutable somehow
+        return self._constrained_director_idx
+
+    @abstractmethod
+    def constrain_values(self, rod: Type[RodBase], time: float) -> None:
+        # TODO: In the future, we can remove rod and use self.system
         """
         Constrain values (position and/or directors) of a rod object.
 
-        In FreeRod class, this routine simply passes.
-
         Parameters
         ----------
         rod : object
             Rod-like object.
         time : float
             The time of simulation.
-
-        Returns
-        -------
-
         """
         pass
 
-    def constrain_rates(self, rod, time):
+    @abstractmethod
+    def constrain_rates(self, rod: Type[RodBase], time: float) -> None:
```

Comment:
why did we change from `system` to `rod`? We can use BC for rigid bodies as well.

Reply:
We should refactor. All constrain_rates and constrain_values in current version `v0.2.1` said `rod`. I think eventually we should remove passing `rod/rigid-body` from outside, instead just call `get_system`.